module/zfs/arc.c

```diff
@@ -8584,15 +8952,67 @@ l2arc_add_vdev(spa_t *spa, vdev_t *vd)
 	list_create(&adddev->l2ad_buflist, sizeof (arc_buf_hdr_t),
 	    offsetof(arc_buf_hdr_t, b_l2hdr.b_l2node));
 
+	list_create(&adddev->l2ad_lbptr_list, sizeof (l2arc_lb_ptr_buf_t),
+	    offsetof(l2arc_lb_ptr_buf_t, node));
+
 	vdev_space_update(vd, 0, 0, adddev->l2ad_end - adddev->l2ad_hand);
 	zfs_refcount_create(&adddev->l2ad_alloc);
+	zfs_refcount_create(&adddev->l2ad_log_blk_count);
 
 	/*
-	 * Add device to global list
+	 * The L2ARC has to hold at least the payload of one log block for
+	 * them to be restored (persistent L2ARC). The payload of a log block
+	 * depends on the amount of its log entries. We always write log blocks
+	 * with 1023 entries. How many of them are committed or restored depends
+	 * on the size of the L2ARC device. Thus the maximum payload of
+	 * one log block is 1023 * SPA_MAXBLOCKSIZE = 16GB. If the L2ARC device
+	 * is less than that, we reduce the amount of committed and restored
+	 * log entries per block so as to enable persistence.
+	 */
+	if (adddev->l2ad_end < l2arc_rebuild_blocks_min_l2size) {
+		adddev->l2ad_log_entries = 0;
+	} else {
+		adddev->l2ad_log_entries = MIN((adddev->l2ad_end -
+		    adddev->l2ad_start) >> SPA_MAXBLOCKSHIFT,
+		    L2ARC_LOG_BLK_MAX_ENTRIES);
+	}
```

Comment:
I think this mean that any device which is smaller than `l2arc_rebuild_blocks_min_l2size` won't write log blocks. Does the code handle a mix of persistent l2arc devices with non-persistent l2arc devices? For example, what if you set `l2arc_rebuild_blocks_min_l2size` to 1TB and add 512GB cache device, then reset `l2arc_rebuild_blocks_min_l2size` back to 1GB and add another 512GB device. Now one device would have `l2ad_log_entries` set to 0 and another which is set to 1023.

Reply:
Good point. I think the code can differentiate, since each device is handled separately. I will add a test case.

Reply:
@grwilson I see now what you mean, I misinterpreted your comment. `l2arc_rebuild_blocks_min_l2size` is defined as a module parameter, it applies to every L2ARC device when `l2arc_add_vdev()` is run. 
So in your scenario, when `l2arc_rebuild_blocks_min_l2size` is set back to 1GB then the old 512GB device would still have no log blocks written in it unless it is offlined/onlined or the pool is exported/imported. The new 512GB will have log blocks written to it because when it was added `l2arc_rebuild_blocks_min_l2size` was already at 1GB.
Given this behaviour, what would you suggest?